File: native/ios/AngularMobile/Sources/AngularMobileRuntime.swift
import Foundation
import WebKit
import UIKit

/// Angular Mobile Runtime
///
/// Main entry point for the Angular Platform Mobile native runtime on iOS.
/// Hosts the JavaScript bundle and provides the bridge to native views.
public class AngularMobileRuntime: NSObject {

    // MARK: - Properties

    private var webView: WKWebView!
    private let viewRegistry = ViewRegistry()
    private let viewFactory: ViewFactory
    private let eventDispatcher = EventDispatcher()
    private let config: RuntimeConfig

    private weak var rootContainer: UIView?
    private var isReady = false
    private var pendingMessages: [String] = []

    // MARK: - Initialization

    public init(config: RuntimeConfig = RuntimeConfig()) {
        self.config = config
        self.viewFactory = ViewFactory(viewRegistry: viewRegistry)
        super.init()
        setupWebView()
    }

    private func setupWebView() {
        let configuration = WKWebViewConfiguration()
        let userContentController = WKUserContentController()

        // Add message handler for bridge communication
        userContentController.add(self, name: "nativeBridge")
        configuration.userContentController = userContentController

        // Enable inline media playback
        configuration.allowsInlineMediaPlayback = true
        configuration.mediaTypesRequiringUserActionForPlayback = []

        webView = WKWebView(frame: .zero, configuration: configuration)
        webView.navigationDelegate = self

        if config.debug {
            if #available(iOS 16.4, *) {
                webView.isInspectable = true
            }
        }
    }

    // MARK: - Public API

    /// Initialize the runtime with a root container
    public func initialize(rootContainer: UIView) {
        self.rootContainer = rootContainer
        viewRegistry.setRootContainer(rootContainer)

        if config.debug {
            print("[AngularMobile] Runtime initialized")
        }
    }

    /// Load the JavaScript bundle from a URL
    public func loadBundle(url: URL) {
        let request = URLRequest(url: url)
        webView.load(request)
    }

    /// Load the JavaScript bundle from the app bundle
    public func loadBundleFromBundle(filename: String = "index", extension ext: String = "html") {
        if let url = Bundle.main.url(forResource: filename, withExtension: ext, subdirectory: "bundle") {
            webView.loadFileURL(url, allowingReadAccessTo: url.deletingLastPathComponent())
        } else {
            print("[AngularMobile] Error: Bundle not found")
        }
    }

    /// Send a message to the JavaScript side
    public func sendToJS(_ message: String) {
        guard isReady else {
            pendingMessages.append(message)
            return
        }

        let escapedMessage = message
            .replacingOccurrences(of: "\\", with: "\\\\")
            .replacingOccurrences(of: "\"", with: "\\\"")

        let script = "window.__handleNativeMessage && window.__handleNativeMessage(\"\(escapedMessage)\")"

        DispatchQueue.main.async { [weak self] in
            self?.webView.evaluateJavaScript(script) { result, error in
                if let error = error, self?.config.debug == true {
                    print("[AngularMobile] JS error: \(error)")
                }
            }
        }
    }

    /// Dispatch an event to JavaScript
    public func dispatchEvent(viewId: String, eventType: String, payload: [String: Any]) {
        let event: [String: Any] = [
            "type": "viewEvent",
            "payload": [
                "viewId": viewId,
                "eventType": eventType,
                "payload": payload
            ]
        ]

        if let jsonData = try? JSONSerialization.data(withJSONObject: event),
           let jsonString = String(data: jsonData, encoding: .utf8) {
            sendToJS(jsonString)
        }
    }

    /// Clean up resources
    public func destroy() {
        webView.configuration.userContentController.removeScriptMessageHandler(forName: "nativeBridge")
        viewRegistry.clear()
        eventDispatcher.clear()
    }

    // MARK: - Message Handling

    private func handleMessage(_ message: [String: Any]) {
        guard let type = message["type"] as? String else { return }

        let id = message["id"] as? String
        let payload = message["payload"] as? [String: Any] ?? [:]

        if config.debug {
            print("[AngularMobile] Received: \(type)")
        }

        do {
            switch type {
            // View Operations
            case "createView":
                try handleCreateView(id: id, payload: payload)
            case "updateView":
                try handleUpdateView(id: id, payload: payload)
            case "removeView":
                try handleRemoveView(id: id, payload: payload)
            case "appendChild":
                try handleAppendChild(id: id, payload: payload)
            case "insertChild":
                try handleInsertChild(id: id, payload: payload)
            case "removeChild":
                try handleRemoveChild(id: id, payload: payload)
            case "setRootView":
                try handleSetRootView(id: id, payload: payload)

            // Batch Operations
            case "batch", "batchOperations":
                try handleBatch(id: id, payload: payload)

            // Measurements
            case "measureView":
                try handleMeasureView(id: id, payload: payload)

            // Focus
            case "focus":
                try handleFocus(id: id, payload: payload)
            case "blur":
                try handleBlur(id: id, payload: payload)

            default:
                print("[AngularMobile] Unknown message type: \(type)")
                sendError(id: id, error: "Unknown message type: \(type)")
            }
        } catch {
            print("[AngularMobile] Error: \(error)")
            sendError(id: id, error: error.localizedDescription)
        }
    }

    // MARK: - View Operations

    private func handleCreateView(id: String?, payload: [String: Any]) throws {
        guard let viewId = payload["viewId"] as? String,
              let viewType = payload["viewType"] as? String else {
            throw RuntimeError.invalidPayload
        }

        let props = payload["props"] as? [String: Any] ?? [:]

        if let view = viewFactory.createView(type: viewType, viewId: viewId, props: props) {
            viewRegistry.register(viewId: viewId, view: view)
            sendSuccess(id: id, data: ["viewId": viewId])
        } else {
            throw RuntimeError.viewCreationFailed(viewType)
        }
    }

    private func handleUpdateView(id: String?, payload: [String: Any]) throws {
        guard let viewId = payload["viewId"] as? String else {
            throw RuntimeError.invalidPayload
        }

        let props = payload["props"] as? [String: Any] ?? [:]

        if let view = viewRegistry.get(viewId: viewId) {
            viewFactory.updateView(view, props: props)
            sendSuccess(id: id)
        } else {
            throw RuntimeError.viewNotFound(viewId)
        }
    }

    private func handleRemoveView(id: String?, payload: [String: Any]) throws {
        guard let viewId = payload["viewId"] as? String else {
            throw RuntimeError.invalidPayload
        }

        viewRegistry.unregister(viewId: viewId)
        sendSuccess(id: id)
    }

    private func handleAppendChild(id: String?, payload: [String: Any]) throws {
        guard let parentId = payload["parentId"] as? String,
              let childId = payload["childId"] as? String else {
            throw RuntimeError.invalidPayload
        }

        guard let parent = viewRegistry.get(viewId: parentId),
              let child = viewRegistry.get(viewId: childId) else {
            throw RuntimeError.viewNotFound(parentId)
        }

        parent.addSubview(child)
        sendSuccess(id: id)
    }

    private func handleInsertChild(id: String?, payload: [String: Any]) throws {
        guard let parentId = payload["parentId"] as? String,
              let childId = payload["childId"] as? String,
              let index = payload["index"] as? Int else {
            throw RuntimeError.invalidPayload
        }

        guard let parent = viewRegistry.get(viewId: parentId),
              let child = viewRegistry.get(viewId: childId) else {
            throw RuntimeError.viewNotFound(parentId)
        }

        parent.insertSubview(child, at: index)
        sendSuccess(id: id)
    }

    private func handleRemoveChild(id: String?, payload: [String: Any]) throws {
        guard let childId = payload["childId"] as? String else {
            throw RuntimeError.invalidPayload
        }

        if let child = viewRegistry.get(viewId: childId) {
            child.removeFromSuperview()
        }
        sendSuccess(id: id)
    }

    private func handleSetRootView(id: String?, payload: [String: Any]) throws {
        guard let viewId = payload["viewId"] as? String else {
            throw RuntimeError.invalidPayload
        }

        if let view = viewRegistry.get(viewId: viewId),
           let root = rootContainer {
            root.subviews.forEach { $0.removeFromSuperview() }
            root.addSubview(view)

            // Fill the container
            view.translatesAutoresizingMaskIntoConstraints = false
            NSLayoutConstraint.activate([
                view.topAnchor.constraint(equalTo: root.topAnchor),
                view.bottomAnchor.constraint(equalTo: root.bottomAnchor),
                view.leadingAnchor.constraint(equalTo: root.leadingAnchor),
                view.trailingAnchor.constraint(equalTo: root.trailingAnchor)
            ])

            sendSuccess(id: id)
        } else {
            throw RuntimeError.viewNotFound(viewId)
        }
    }

    private func handleBatch(id: String?, payload: [String: Any]) throws {
        let operations = payload["operations"] as? [[String: Any]] ?? payload["messages"] as? [[String: Any]] ?? []

        for op in operations {
            handleMessage(op)
        }

        sendSuccess(id: id)
    }

    private func handleMeasureView(id: String?, payload: [String: Any]) throws {
        guard let viewId = payload["viewId"] as? String else {
            throw RuntimeError.invalidPayload
        }

        if let view = viewRegistry.get(viewId: viewId),
           let window = view.window {
            let frame = view.convert(view.bounds, to: window)

            sendSuccess(id: id, data: [
                "x": frame.origin.x,
                "y": frame.origin.y,
                "width": frame.width,
                "height": frame.height
            ])
        } else {
            throw RuntimeError.viewNotFound(viewId)
        }
    }

    private func handleFocus(id: String?, payload: [String: Any]) throws {
        guard let viewId = payload["viewId"] as? String else {
            throw RuntimeError.invalidPayload
        }

        if let view = viewRegistry.get(viewId: viewId) {
            view.becomeFirstResponder()
        }
        sendSuccess(id: id)
    }

    private func handleBlur(id: String?, payload: [String: Any]) throws {
        guard let viewId = payload["viewId"] as? String else {
            throw RuntimeError.invalidPayload
        }

        if let view = viewRegistry.get(viewId: viewId) {
            view.resignFirstResponder()
        }
        sendSuccess(id: id)
    }

    // MARK: - Response Helpers

    private func sendSuccess(id: String?, data: [String: Any]? = nil) {
        var response: [String: Any] = [
            "id": id ?? "",
            "success": true
        ]
        if let data = data {
            response["data"] = data
        }

        if let jsonData = try? JSONSerialization.data(withJSONObject: response),
           let jsonString = String(data: jsonData, encoding: .utf8) {
            sendToJS(jsonString)
        }
    }

    private func sendError(id: String?, error: String) {
        let response: [String: Any] = [
            "id": id ?? "",
            "success": false,
            "error": error
        ]

        if let jsonData = try? JSONSerialization.data(withJSONObject: response),
           let jsonString = String(data: jsonData, encoding: .utf8) {
            sendToJS(jsonString)
        }
    }
}

// MARK: - WKScriptMessageHandler

extension AngularMobileRuntime: WKScriptMessageHandler {
    public func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        guard message.name == "nativeBridge" else { return }

        if let messageString = message.body as? String,
           let data = messageString.data(using: .utf8),
           let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            handleMessage(json)
        } else if let json = message.body as? [String: Any] {
            handleMessage(json)
        }
    }
}

// MARK: - WKNavigationDelegate

extension AngularMobileRuntime: WKNavigationDelegate {
    public func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        isReady = true

        if config.debug {
            print("[AngularMobile] Bundle loaded")
        }

        // Send any pending messages
        pendingMessages.forEach { sendToJS($0) }
        pendingMessages.removeAll()
    }

    public func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
        print("[AngularMobile] Navigation failed: \(error)")
    }
}

// MARK: - Supporting Types

public struct RuntimeConfig {
    public var debug: Bool
    public var hotReload: Bool
    public var bundleUrl: URL?

    public init(debug: Bool = false, hotReload: Bool = false, bundleUrl: URL? = nil) {
        self.debug = debug
        self.hotReload = hotReload
        self.bundleUrl = bundleUrl
    }
}

enum RuntimeError: LocalizedError {
    case invalidPayload
    case viewNotFound(String)
    case viewCreationFailed(String)

    var errorDescription: String? {
        switch self {
        case .invalidPayload:
            return "Invalid message payload"
        case .viewNotFound(let viewId):
            return "View not found: \(viewId)"
        case .viewCreationFailed(let viewType):
            return "Failed to create view of type: \(viewType)"
        }
    }
}

File: native/android/runtime/src/main/kotlin/com/pegasusheavy/angularmobile/AngularMobileRuntime.kt
package com.pegasusheavy.angularmobile

import android.content.Context
import android.util.Log
import android.webkit.JavascriptInterface
import android.webkit.WebView
import android.webkit.WebViewClient
import android.webkit.WebSettings
import org.json.JSONObject
import org.json.JSONArray

/**
 * Angular Mobile Runtime
 *
 * Main entry point for the Angular Platform Mobile native runtime.
 * Hosts the JavaScript bundle and provides the bridge to native views.
 */
class AngularMobileRuntime(
    private val context: Context,
    private val config: RuntimeConfig = RuntimeConfig()
) {
    companion object {
        private const val TAG = "AngularMobile"
        private const val BRIDGE_NAME = "__ANDROID_BRIDGE__"
    }

    private lateinit var webView: WebView
    private val viewRegistry = ViewRegistry()
    private val viewFactory = ViewFactory(context, viewRegistry)
    private val eventDispatcher = EventDispatcher()

    private var jsCallback: ((String) -> Unit)? = null
    private var isReady = false

    /**
     * Initialize the runtime with a root container
     */
    fun initialize(rootContainer: android.view.ViewGroup) {
        viewRegistry.setRootContainer(rootContainer)
        setupWebView()

        if (config.debug) {
            Log.d(TAG, "Angular Mobile Runtime initialized")
        }
    }

    /**
     * Load the JavaScript bundle
     */
    fun loadBundle(bundleUrl: String) {
        webView.loadUrl(bundleUrl)
    }

    /**
     * Load the JavaScript bundle from assets
     */
    fun loadBundleFromAssets(assetPath: String = "bundle/index.html") {
        webView.loadUrl("file:///android_asset/$assetPath")
    }

    private fun setupWebView() {
        webView = WebView(context).apply {
            settings.apply {
                javaScriptEnabled = true
                domStorageEnabled = true
                allowFileAccess = true
                allowContentAccess = true
                mediaPlaybackRequiresUserGesture = false

                if (config.debug) {
                    WebView.setWebContentsDebuggingEnabled(true)
                }
            }

            webViewClient = object : WebViewClient() {
                override fun onPageFinished(view: WebView?, url: String?) {
                    super.onPageFinished(view, url)
                    isReady = true
                    if (config.debug) {
                        Log.d(TAG, "Bundle loaded: $url")
                    }
                }
            }

            // Add the native bridge
            addJavascriptInterface(NativeBridge(), BRIDGE_NAME)
        }
    }

    /**
     * Send a message to the JavaScript side
     */
    fun sendToJS(message: String) {
        if (!isReady) {
            Log.w(TAG, "Runtime not ready, message queued")
            return
        }

        val escapedMessage = message.replace("\\", "\\\\").replace("\"", "\\\"")
        val script = "window.__handleNativeMessage && window.__handleNativeMessage(\"$escapedMessage\")"

        webView.post {
            webView.evaluateJavascript(script) { result ->
                if (config.debug) {
                    Log.d(TAG, "JS response: $result")
                }
            }
        }
    }

    /**
     * Native bridge exposed to JavaScript
     */
    inner class NativeBridge {

        @JavascriptInterface
        fun postMessage(messageJson: String) {
            try {
                val message = JSONObject(messageJson)
                handleMessage(message)
            } catch (e: Exception) {
                Log.e(TAG, "Failed to parse message: $messageJson", e)
                sendError(null, "Failed to parse message: ${e.message}")
            }
        }

        @JavascriptInterface
        fun registerCallback(callbackName: String) {
            // JavaScript will call this to register for callbacks
            if (config.debug) {
                Log.d(TAG, "Callback registered: $callbackName")
            }
        }
    }

    /**
     * Handle incoming message from JavaScript
     */
    private fun handleMessage(message: JSONObject) {
        val type = message.optString("type")
        val id = message.optString("id")
        val payload = message.optJSONObject("payload") ?: JSONObject()

        if (config.debug) {
            Log.d(TAG, "Received: $type")
        }

        try {
            when (type) {
                // View Operations
                "createView" -> handleCreateView(id, payload)
                "updateView" -> handleUpdateView(id, payload)
                "removeView" -> handleRemoveView(id, payload)
                "appendChild" -> handleAppendChild(id, payload)
                "insertChild" -> handleInsertChild(id, payload)
                "removeChild" -> handleRemoveChild(id, payload)
                "setRootView" -> handleSetRootView(id, payload)

                // Batch Operations
                "batch", "batchOperations" -> handleBatch(id, payload)

                // Measurements
                "measureView" -> handleMeasureView(id, payload)

                // Focus
                "focus" -> handleFocus(id, payload)
                "blur" -> handleBlur(id, payload)

                // Events
                "addEventListener" -> handleAddEventListener(id, payload)
                "removeEventListener" -> handleRemoveEventListener(id, payload)

                else -> {
                    Log.w(TAG, "Unknown message type: $type")
                    sendError(id, "Unknown message type: $type")
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error handling message: $type", e)
            sendError(id, e.message ?: "Unknown error")
        }
    }

    // ==================== View Operations ====================

    private fun handleCreateView(messageId: String, payload: JSONObject) {
        val viewId = payload.getString("viewId")
        val viewType = payload.getString("viewType")
        val props = payload.optJSONObject("props") ?: JSONObject()

        val view = viewFactory.createView(viewType, viewId, props)

        if (view != null) {
            viewRegistry.register(viewId, view)
            sendSuccess(messageId, JSONObject().put("viewId", viewId))
        } else {
            sendError(messageId, "Failed to create view of type: $viewType")
        }
    }

    private fun handleUpdateView(messageId: String, payload: JSONObject) {
        val viewId = payload.getString("viewId")
        val props = payload.optJSONObject("props") ?: JSONObject()

        val view = viewRegistry.get(viewId)
        if (view != null) {
            viewFactory.updateView(view, props)
            sendSuccess(messageId)
        } else {
            sendError(messageId, "View not found: $viewId")
        }
    }

    private fun handleRemoveView(messageId: String, payload: JSONObject) {
        val viewId = payload.getString("viewId")

        viewRegistry.unregister(viewId)
        sendSuccess(messageId)
    }

    private fun handleAppendChild(messageId: String, payload: JSONObject) {
        val parentId = payload.getString("parentId")
        val childId = payload.getString("childId")

        val parent = viewRegistry.get(parentId) as? android.view.ViewGroup
        val child = viewRegistry.get(childId)

        if (parent != null && child != null) {
            parent.addView(child)
            sendSuccess(messageId)
        } else {
            sendError(messageId, "Parent or child not found")
        }
    }

    private fun handleInsertChild(messageId: String, payload: JSONObject) {
        val parentId = payload.getString("parentId")
        val childId = payload.getString("childId")
        val index = payload.getInt("index")

        val parent = viewRegistry.get(parentId) as? android.view.ViewGroup
        val child = viewRegistry.get(childId)

        if (parent != null && child != null) {
            parent.addView(child, index)
            sendSuccess(messageId)
        } else {
            sendError(messageId, "Parent or child not found")
        }
    }

    private fun handleRemoveChild(messageId: String, payload: JSONObject) {
        val parentId = payload.getString("parentId")
        val childId = payload.getString("childId")

        val parent = viewRegistry.get(parentId) as? android.view.ViewGroup
        val child = viewRegistry.get(childId)

        if (parent != null && child != null) {
            parent.removeView(child)
            sendSuccess(messageId)
        } else {
            sendError(messageId, "Parent or child not found")
        }
    }

    private fun handleSetRootView(messageId: String, payload: JSONObject) {
        val viewId = payload.getString("viewId")
        val view = viewRegistry.get(viewId)

        if (view != null) {
            viewRegistry.getRootContainer()?.let { root ->
                root.removeAllViews()
                root.addView(view)
            }
            sendSuccess(messageId)
        } else {
            sendError(messageId, "View not found: $viewId")
        }
    }

    private fun handleBatch(messageId: String, payload: JSONObject) {
        val operations = payload.optJSONArray("operations")
            ?: payload.optJSONArray("messages")
            ?: JSONArray()

        for (i in 0 until operations.length()) {
            val op = operations.getJSONObject(i)
            handleMessage(op)
        }

        sendSuccess(messageId)
    }

    private fun handleMeasureView(messageId: String, payload: JSONObject) {
        val viewId = payload.getString("viewId")
        val view = viewRegistry.get(viewId)

        if (view != null) {
            val location = IntArray(2)
            view.getLocationOnScreen(location)

            val result = JSONObject().apply {
                put("x", location[0])
                put("y", location[1])
                put("width", view.width)
                put("height", view.height)
            }
            sendSuccess(messageId, result)
        } else {
            sendError(messageId, "View not found: $viewId")
        }
    }

    private fun handleFocus(messageId: String, payload: JSONObject) {
        val viewId = payload.getString("viewId")
        val view = viewRegistry.get(viewId)

        view?.requestFocus()
        sendSuccess(messageId)
    }

    private fun handleBlur(messageId: String, payload: JSONObject) {
        val viewId = payload.getString("viewId")
        val view = viewRegistry.get(viewId)

        view?.clearFocus()
        sendSuccess(messageId)
    }

    private fun handleAddEventListener(messageId: String, payload: JSONObject) {
        val viewId = payload.getString("viewId")
        val eventType = payload.getString("eventType")

        eventDispatcher.register(viewId, eventType) { eventData ->
            dispatchEvent(viewId, eventType, eventData)
        }
        sendSuccess(messageId)
    }

    private fun handleRemoveEventListener(messageId: String, payload: JSONObject) {
        val viewId = payload.getString("viewId")
        val eventType = payload.getString("eventType")

        eventDispatcher.unregister(viewId, eventType)
        sendSuccess(messageId)
    }

    // ==================== Response Helpers ====================

    private fun sendSuccess(messageId: String?, data: JSONObject? = null) {
        val response = JSONObject().apply {
            put("id", messageId)
            put("success", true)
            if (data != null) {
                put("data", data)
            }
        }
        sendToJS(response.toString())
    }

    private fun sendError(messageId: String?, error: String) {
        val response = JSONObject().apply {
            put("id", messageId)
            put("success", false)
            put("error", error)
        }
        sendToJS(response.toString())
    }

    /**
     * Dispatch an event to JavaScript
     */
    fun dispatchEvent(viewId: String, eventType: String, payload: JSONObject) {
        val event = JSONObject().apply {
            put("type", "viewEvent")
            put("payload", JSONObject().apply {
                put("viewId", viewId)
                put("eventType", eventType)
                put("payload", payload)
            })
        }
        sendToJS(event.toString())
    }

    /**
     * Clean up resources
     */
    fun destroy() {
        webView.destroy()
        viewRegistry.clear()
        eventDispatcher.clear()
    }
}

/**
 * Runtime configuration
 */
data class RuntimeConfig(
    val debug: Boolean = false,
    val hotReload: Boolean = false,
    val bundleUrl: String? = null
)
